Working in-progress v3

halx99 opened the pull request:
## Describe your changes

Previous working PR  #2204 was closed due to branch name renamed from `3.x` to `wip-v3`

- [x] ISO c++23 standard
- [x] Texture resource improvment, add texture array support, remove CPU array implement since all graphics API support texture array
- [x] Remove sampler from Texture, implement texture sampler object and cache it, since GLES3/OGL-3.3, mtl, d3d support sampler object
- [x] Update lua to 5.5.x
- [x] Remove the default pixel format, fix rendering of grayscale PNGs by @j-jorge in https://github.com/axmolengine/axmol/pull/2573
- [x] Remove all deprecated stubs
- [x] Implement drawArraysInstanced and make instanced draw more common use
- [x] box2d-3.x
    - [x] Make Box2dTest of `cpp-tests` works on box2d v3
    - [x] Make Box2dTestBed of `cpp-tests` works on box2d v3
- [x] Remove chipmunk
    - [x] Remove `core/physics` based on chipmunk
    - [x] Remove ChipmunkTest from `cpp-tests`
    - [x] Remove ChipmuniTestbed from `cpp-tests`
- [x] Improve DrawNode, don't create backend::Buffer always
- [x] Upgrade iOS deploy target: 12.0 => 13.0
- [x] Upgrade android minsdk: 17(Android 4.2) => 21(Android 5.1)
- [x] Upgrade android ndk: r23 => latest LTS release r27
- [x] Due to android,ios toolchain update to support c++20 full features, `std::views::split` can be use to improve performance
- [x] Remove  3rdparty ghc filesystem, since both ios13 and android5.1 support `std::filesystem`
- [x] Build and run wasm64 support, current need enable chrome flag: `chrome://flags/#enable-webassembly-memory64`
- [x] `Color4B` => `Color32`
- [x] Remove `Color3B`
- [x] Remove deprecated StringUtils::format
- [x] drop GLES 2.0 support since all Android 5.1+ devices support `opengl-es-3.0`
- [x] Improve axslcc preprocessors, remove `METAL`, `GLES2`, then add `AXSLC_TARGET_[GLSL,HLSL,MSL,GLES]`
- [x] Improve VertexLayout management
- [x] Improve `sampler2D texs[4]` support, 2.x only support OpenGL, now support all RHI: d3d11, metal, opengl(es)
- [ ] Reimplement core/physics with box2d v3, not all API compatible with axmol-2.x (in-progress)
     - [x] basic functionals 
     - [x] Rename `PhysicsJointXXX::construct` => `PhysicsJointXXX::instantiate`
     - [ ] b2ContactListener
- [ ] re-select json lib: yyjson > boost.json > rapidjson
    simdjson for parse only and simdjson build in-progress: https://github.com/simdjson/experimental_json_builder
    finally solution: `yyjson`  + `simdjson`
- [ ] drop tolua, use sol2
- [ ] Choose shader for more render objects like `Sprite`
    - [x] ParticleSystemQuad

## Issue ticket number and link


## Checklist before requesting a review
### For each PR
- [ ] Add Copyright if it missed:   
      - `"Copyright (c) 2019-present Axmol Engine contributors (see AUTHORS.md)."`
- [ ] I have performed a self-review of my code.
       
   Optional:
   - [ ] I have checked readme and add important infos to this PR.
   - [ ] I have added/adapted some tests too.
          
### For core/new feature PR
- [ ] I have checked readme and add important infos to this PR.
- [ ] I have added thorough tests.

halx99 committed:
Remove deprecated stubs

halx99 committed:
Implement drawArraysInstanced and make instanced draw more common use

halx99 committed:
Bump 3.0.0

halx99 committed:
Mark DrawNode::clear as virtual function

halx99 committed:
Make box2d 3.0 works and remove all chipmunk related

halx99 committed:
Change underlaying color type to Color4F aka ax::Color

halx99 committed:
Improve DrawNode, don't create backend::Buffer always

halx99 committed:
Fix ci

halx99 committed:
Improve code style

halx99 committed:
Fix assetion fail in drawOrder enabled

halx99 committed:
Make DrawNodeTest CandyMixEffect behavior same with 2.x

halx99 committed:
Tidy vertex format

halx99 committed:
All typo: ax::Color4F ==> ax::Color

halx99 committed:
Fix LayerGradient not adapte float4 color format

halx99 committed:
Delete unused files

halx99 committed:
Fix preprocessor checks

halx99 committed:
Update openal-soft to 1.24.1

halx99 committed:
Fix #2254

halx99 committed:
Rename all XXX_C4F_T2F_XXX to XXX_T2F_C4F_XXX

halx99 committed:
Still use V3F_T2F_C4B for sprite render

halx99 committed:
Update lua bindings

halx99 committed:
Merge axmol dev into 3.x

rh101 committed:
AxmolActivity will extend AppCompatActivity
Min SDK set to 21
Tests updated to support AppCompatActivity
MultiDexEnabled for cpp-tests

halx99 committed:
Update Android & iOS minmal OS support

Android: Android 5.1(API 21)+
iOS: 13+

halx99 committed:
Update simdjson to v3.11.4

halx99 committed:
Remove dependency rapidjson for JsonWriter

halx99 committed:
Update openal-soft to 1.24.2

halx99 committed:
Update simdjson to v3.11.5

halx99 committed:
Fix #2330 and improve code style

halx99 committed:
Improve code style

halx99 committed:
Fix linux link error

halx99 committed:
Rename Color4B to Color32

halx99 committed:
Fix wasm64 build

halx99 committed:
Fix wasm64 runtime exception for imgui glfw backend

halx99 committed:
Add wasm64 build to ci

halx99 committed:
Improve Console::Utility::trim and move to StringUtils

    - Use c++20 reature ranges
    - Use c++20 0-GC split to instead std::getline

halx99 committed:
Add lib: fast_float

-  Resolve clang libc++ not support std::from_chars for floating points

halx99 committed:
Improve c3t, c3b load

- Use simdjson to load c3t
- Use yasio::ibstream to c3b

halx99 committed:
Fix android appcompat not set in template project

halx99 committed:
Fix wasm compiling warnigs and errors

halx99 committed:
Update simdjson and fast_float

halx99 committed:
Fixup merge issue cause compile errors

paulocoutinhox commented:
Version 3 don't will be published with **PhysicsBody** replacement?

<img width="1507" alt="image" src="https://github.com/user-attachments/assets/b106a56d-2af0-417f-8af5-d6a048558b89" />

halx99 commented:
version 3 still in early status, and should contains physics2d when publish.

paulocoutinhox commented:
Will be nice if we have a single json library easy to use, like `nlohmann/json` instead of two.

halx99 commented:
> Will be nice if we have a single json library easy to use, like `nlohmann/json` instead of two.

The `nlohmann/json` easy to use, but performance not good.

danialias commented:
I need rapidjson performance as well, please do not remove 😅

paulocoutinhox commented:
I don't understand how a JSON lib in a game engine needs such an absurd performance, since things like that are used in few moments, like loading or saving things at specific times.

In our app (Ubook) that uses JSON all the time and processes thousands of JSON data with thousands of records, the processing is instantaneous, even in huge JSON data and we use nloman/json. Before we used rapid, but it doesn't have as many features, requiring other libs to be used and more things to be maintained, decreasing efficiency in exchange for NOTHING.

For me, it would make sense to keep a single lib that is easy to use and that for our scenario here at Axmol, will not affect anything in practice.

Here is my opinion based on my reality as someone who uses this in practice and in large volumes, both on Android and iOS, as well as on Apple Watch and Apple TV and also on the desktop.

rh101 commented:
> I need rapidjson performance as well, please do not remove 😅

If it is removed from Axmol, then you can easily add it into your own projects with a few lines of CMake code, and nothing else has to change in your projects.

rh101 commented:
> I don't understand how a JSON lib in a game engine needs such an absurd performance, since things like that are used in few moments, like loading or saving things at specific times.

Your projects may not require the speed of libraries like rapidjson, but that may not apply to projects by others, since they may have completely different requirements.

> In our app (Ubook) that uses JSON all the time and processes thousands of JSON data with thousands of records, the processing is instantaneous, even in huge JSON data and we use nloman/json. Before we used rapid, but it doesn't have as many features, requiring other libs to be used and more things to be maintained, decreasing efficiency in exchange for NOTHING.
 
There are cases where the speed of parsing needs to be as quick as possible, such as in event handlers, comms response handlers, etc. etc..  In these cases, speed definitely does matter, and it is quite noticeable if the parsing takes too long.

> For me, it would make sense to keep a single lib that is easy to use and that for our scenario here at Axmol, will not affect anything in practice.

That may be the case for your own projects, but again, it may not apply to projects by others.  Also, just because a specific library works for you, it doesn't mean that it is appropriate for it to be used in Axmol.

> Here is my opinion based on my reality as someone who uses this in practice and in large volumes, both on Android and iOS, as well as on Apple Watch and Apple TV and also on the desktop.

That's all well and good, and as you stated, your opinion is based on the requirements of your own projects, but it's not really appropriate to make statements about what other developer should or shouldn't use in their projects, because you really don't know what their requirements are.  

The engine should bundle whatever json library (or libraries) it requires, but that in no way forces the developers using Axmol to also use those libraries.  Developers are free to use whatever library they want to use.

On a side note, many developers who use Axmol for their released projects have never posted on this discussion board, and may not even read these threads, so unfortunately, it would be hard to get their feedback on such things.

halx99 commented:
Finally, maybe `yyjson`, speed fast and support read/write and dom style APIs, axmol can make a thin  c++ wrapper

iAndyHD3 commented:
The problem with nlohman/json is not only its lack of runtime speed but it's also a huge >10k thousand long header only library, making compilation also slower. its really one of the worst options you can choose. In modern C++ there is a clear new winner in my opinion:  [glaze](https://github.com/stephenberry/glaze) is a modern C++ json library with automatic serialization/deserialization, making it really convenient for simple use cases. The performance is on par with simdjson. For non reflection uses, there is [generic json](https://github.com/stephenberry/glaze/blob/main/docs/generic-json.md) which is similar to nlohman in that regard.

There are a few cons, thought:
it only supports c++23
Needs recent compiler versions
Not sure if it has been tested in platforms like wasm, ios

paulocoutinhox commented:
Hi,

Glaze is nice too. Their syntax is like nlohman/json and is easy to use.

My concern is always to keep the engine as easy as possible for people to use. It's better to use Glaze, which already does everything, than to have to adopt several libraries and increase complexity, lowering the barrier to entry.

halx99 commented:
I tried glaze, but compile fails
EDIT: seems it fix msvc compile issues recently

halx99 committed:
[WIP] Implement Physics2D with box2d-v3

halx99 committed:
Improve type Color32

WUCJ638 commented:
I hope DragonBones extensions is retained, for it eases my creating skeleton animation, and cares for those without access to Spine PRO like me, though it's no more under maintenance.

Yet it's said that currently a NEW DragonBone (aka "Loongbones") has been published and under developement, so I believe it's still reasonable to keep it around.🤔

rh101 commented:
> I hope DragonBones extensions is retained, for it eases my creating skeleton animation, and cares for those without access to Spine PRO like me, though it's no more under maintenance.
> 
> Yet it's said that currently a NEW DragonBone (aka "Loongbones") has been published and under developement, so I believe it's still reasonable to keep it around.🤔

You can easily add it directly to your game project, since the engine doesn't actually interact with DragonBones, so it does not need to exist in the engine source tree.

WUCJ638 commented:
> You can easily add it directly to your game project, since the engine doesn't actually interact with DragonBones, so it does not need to exist in the engine source tree.

That is I manually add it to my project or the Engine in my computer...🤔acceptable. Thank you😊.

halx99 commented:
actually, we can reserve all exists axmol  extensions

halx99 committed:
Improve FileUtils removeDirectory implementation, see #2208

halx99 committed:
Update openal-soft to 1.24.3

WUCJ638 commented:
> actually, we can reserve all exists axmol extensions

Great, enabling me to code smoothly. Thanks!😉

aismann commented:
@halx99 
Just a suggestion:
About: _Renaming Color4B to Color32_

Isn't `Color256` more self-explanatory and therefore better?

halx99 commented:
> @halx99 Just a suggestion: About: _Renaming Color4B to Color32_
> 
> Isn't `Color256` more self-explanatory and therefore better?

Why 256， it's 32bit

rh101 commented:
> Isn't `Color256` more self-explanatory and therefore better?

Color256 does not give you any information at all on the contents of that structure. It really makes no sense at all, given that the actual size is 2^32, not 2^8.  

The only naming that works is either ColorXX, where XX is the number of bits, or as it is right now, ColorXY, where X is the number of values, and Y is the type, such as Color4B/Color4F etc. etc..

aismann commented:
@halx99 @rh101 
Thanks for explanations. 

I thought Color256 because of the used values 0-255.
Color32 is ok for me too.

aismann commented:
[Box2D Releases](https://github.com/erincatto/box2d/releases) [v3.1.0](https://github.com/erincatto/box2d/releases/tag/v3.1.0)

halx99 committed:
Update codebase to c++23

halx99 committed:
Fix possibile conflict

aismann commented:
@halx99 
[Box2D Release v3.1.1 · erincatto/box2d](https://github.com/erincatto/box2d/releases/tag/v3.1.1)
This is a bug fix release. There are a few minor API changes. Making way for 3.2.0 work.
There is no documentation updates with this releases.

What's Changed

- Warning as error off by default by @erincatto in https://github.com/erincatto/box2d/pull/914
- functions to get and set surface material by @erincatto in https://github.com/erincatto/box2d/pull/921
- Fixed another shape distance assert by @erincatto in https://github.com/erincatto/box2d/pull/926
- Revolute and prismatic targets by @erincatto in https://github.com/erincatto/box2d/pull/929
- Cast initial overlap by @erincatto in https://github.com/erincatto/box2d/pull/933
- Tunable joint stiffness by @erincatto in https://github.com/erincatto/box2d/pull/934
- _aligned_free for mingw by @erincatto in https://github.com/erincatto/box2d/pull/939
- Removed MSVC static library setting by @erincatto in https://github.com/erincatto/box2d/pull/940

Full Changelog: https://github.com/erincatto/box2d/compare/v3.1.0...v3.1.1

halx99 committed:
Update simdjson: 3.12.0 => 3.13.0

halx99 committed:
Update fast_float: 8.0.0 => 8.0.2

halx99 committed:
Fix box2d-v3 shape instancing draw for metal render backend

aismann commented:
@halx99 
`ax::LayerGradient `is not working, see also
```
10::LayerGradientTest
11::LayerGradientTest2
```

halx99 commented:
> @halx99 `ax::LayerGradient `is not working, see also
> 
> ```
> 10::LayerGradientTest
> 11::LayerGradientTest2
> ```

Thanks, will take a look later

aismann commented:
Here a yasio bug (
![image](https://github.com/user-attachments/assets/c674b530-024b-4508-aecd-c1953dca4d03)
I got it running the "Start AutoTest" on cpp-tests

halx99 committed:
Catch Bundle3D binary read exception

halx99 committed:
Fix LayerGradient color

halx99 commented:
> Here a yasio bug ( ![image](https://private-user-images.githubusercontent.com/8652787/456576915-c674b530-024b-4508-aecd-c1953dca4d03.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTAyNjE5NDYsIm5iZiI6MTc1MDI2MTY0NiwicGF0aCI6Ii84NjUyNzg3LzQ1NjU3NjkxNS1jNjc0YjUzMC0wMjRiLTQ1MDgtYWVjZC1jMTk1M2RjYTRkMDMucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDYxOCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTA2MThUMTU0NzI2WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MWRmY2VjMjkwOGI1ZDc3MjUyNzE3MTFiOTJjYzI4YzVjYTQ5MmEwZjRmZjE4M2E1NDNkMmU4Yzg5ZDdjM2ZmMyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.25ufWv95MdU46FeaNkK9x9F4ZGlv7r4qtqTF_171FYI) I got it running the "Start AutoTest" on cpp-tests

fixed

halx99 commented:
> > @halx99 `ax::LayerGradient `is not working, see also
> > ```
> > 10::LayerGradientTest
> > 11::LayerGradientTest2
> > ```
> 
> Thanks, will take a look later

fixed

j-jorge committed:
Remove the default pixel format, fix rendering of grayscale PNGs (#2573)

* Add AXMOL_AUTOTEST_CAPTURE_DIR for cpp-tests.

In auto test mode, cpp-tests will save a capture of the scene in the
directory whose path is given by environment variable
AXMOL_AUTOTEST_CAPTURE_DIR, if such a path is provided.

* Check the value of AXMOL_START_AUTOTEST, ignore it if zero.

* Remove the default pixel format from Texture2D, use the source format.

* Handle gray+alpha PNG images as gray+alpha sprites.

halx99 committed:
Fix merging issues due to imgui updated to 1.92.0

halx99 committed:
Update plainlua to 5.5.0-beta

halx99 committed:
Fix luabindings for physics2d

halx99 committed:
Fix merging issues

halx99 committed:
Add common API to setup programState for sprite shader based Node, finish #2603

j-jorge commented on core/renderer/backend/ProgramManager.cpp:
```diff
@@ -55,6 +55,22 @@ void ProgramManager::destroyInstance()
+    case PixelFormat::RGBA8:
+        return backend::ProgramType::POSITION_TEXTURE_COLOR;
+    default:
+        AXLOGW("Warning: Sprite::chooseSpriteShader() unhandled pixel format {}", (int)pixelFormat);
```
Bad log, it's not `Sprite::chooseSpriteShader()` anymore :)

j-jorge commented:
What do you think of taking advantage of the break introduced by v3 to remove the axmol.h file? This is the kind of header that explodes the build duration :)

halx99 committed:
Fix chooseSpriteProgramType log message typo

halx99 committed:
Implment some physics2d joints, rename PhysicsJointXXX::construct to instantiate

halx99 committed:
Enable physics 2d lua bindings

halx99 committed:
Better lua 5.5 support

halx99 commented:
> What do you think of taking advantage of the break introduced by v3 to remove the axmol.h file? This is the kind of header that explodes the build duration :)

Need consider long term.

aismann commented:
> > What do you think of taking advantage of the break introduced by v3 to remove the axmol.h file? This is the kind of header that explodes the build duration :)
> 
> Need consider long term.

Maybe a text like the "copyright hint" (pull_request_template.md) is a "start" on each PR?

[ ]  Check the 'axmol.h' and replace it with the needed headers.

halx99 commented:
> > > What do you think of taking advantage of the break introduced by v3 to remove the axmol.h file? This is the kind of header that explodes the build duration :)
> > 
> > 
> > Need consider long term.
> 
> Maybe a text like the "copyright hint" (pull_request_template.md) is a "start" on each PR?
> 
> [ ] Check the 'axmol.h' and replace it with the needed headers.

I agree

j-jorge commented:
I'm not sure I understand your feedback correctly, but if we keep the axmol.h file maybe we can put a deprecation warning message in it, and at least remove the `#include <axmol.h>` from this repo? Then we'll be able to remove it in v3.x or v4.

aismann commented:
<img width="211" height="46" alt="image" src="https://github.com/user-attachments/assets/805d72cd-a814-478c-8aa1-bb34f4b7bf6a" />

<img width="1043" height="644" alt="image" src="https://github.com/user-attachments/assets/be062343-2507-4508-ade8-73ab76b0761e" />

aismann commented:
> > > > What do you think of taking advantage of the break introduced by v3 to remove the axmol.h file? This is the kind of header that explodes the build duration :)
> > > 
> > > 
> > > Need consider long term.
> > 
> > 
> > Maybe a text like the "copyright hint" (pull_request_template.md) is a "start" on each PR?
> > [ ] Check the 'axmol.h' and replace it with the needed headers.
> 
> I agree

I will change the `pull_request_template.md`
@halx99 
For the 3.0 branch, right?

aismann commented:
> I'm not sure I understand your feedback correctly, but if we keep the axmol.h file maybe we can put a deprecation warning > message in it, and at least remove the `#include <axmol.h>` from this repo? Then we'll be able to remove it in v3.x or v4.

@j-jorge 
We will start with v3.x.

halx99 committed:
🔥Remove Color3B (#2607)

halx99 committed:
🛠 ️Remove redundant virtual marks (#2610)

halx99 committed:
Update ndk: r27c => r27d

halx99 committed:
🔥 Remove deprecated StringUtils::format (#2613)

halx99 committed:
🛠  Improve code style

halx99 committed:
🛠  Fix lua bindings

halx99 committed:
🛠  Update llvm to 20.1.8 in build.profiles

aismann commented:
@halx99 
There is a missing assign on `void Label::setTextColor(const Color32& color)`
Correct is:
`    _textColor = _textColor32 = color;`

Maybe all `_textColor` have to rename to `_textColor32`?

halx99 committed:
🛠  Fix label color set

halx99 commented:
> @halx99 There is a missing assign on `void Label::setTextColor(const Color32& color)` Correct is: ` _textColor = _textColor32 = color;`
> 
> Maybe all `_textColor` have to rename to `_textColor32`?

fixed by https://github.com/axmolengine/axmol/pull/2387/commits/284c22989e1ca635eb877883e90bf07c5d97816a

halx99 committed:
Fix merging issues

halx99 committed:
Merge LayoutHelper to ui::Helper

halx99 committed:
🛠  Fix ui:Helper lint warnings

halx99 committed:
🔥 Optimize RichText::handleTextRenderer to reduce memory copies using zero-copy techniques

- Improve `text_utils::StringUTF8` and rename to `text_utils::u8char_span`

halx99 committed:
Fix lint issues

halx99 committed:
Fix merging issues

halx99 committed:
Remove deprecated stubs

halx99 committed:
Fix merging issues

halx99 committed:
Implement d3d11 renderer backend (#2631)

* Improve render backend macro detection

* Implement d3d11 backend (WIP)

* [d3d11] Fix depth stencil state not apply

Also resolved self-addressed issue: text rendering "Te", the 'e' left will edge pixel will lost, actually override by "T" due to draw order mistake

* [d3d] Fix buffer creation

* [d3d] Make ImGui works

* [d3d11] Resize SwapChain support

Also fix texture reference leaks

* Improve VertexLayout

* Update axslcc to 1.10.0 to resolve d3d instance draw issue

Reason: the older axslcc split mat4 vertex input four vec4 when translate to HLSL will allocate non-standard semantic name, looks like `TEXCOORD1_[1~N]`, benefit SPIRV-cross support flag: flatten_matrix_vertex_input_semantics=true, SPIRV-corss will translate mat4 to TEXCOORD{N,N+1,N+2,...} rather than TEXCOORDN_{0,1,2,3}

* [d3d11] Fix RGB8 texture upload

The d3d not support RGB8, need convert at renderer frontend

* Resolve lint warnings

* Update luabindings generator

* [d3d11] Make live2d works

* [d3d11] Translate texture descriptor properly

* Fix compile errors

* [d3d11] Try fix depth-stencil texture creation

* [d3d11] Fix offscreen render target depth-stencil texture creation

* [d3d11] Fix OMSetRenderTargets: Resource being set to OM RenderTarget slot 0 is still bound on input

* [d3d11] bind dummy lightmap texture for terrain when lightmap not set by user

* [d3d11] Add mip levels support temporary

* [d3d11] Fix skybox textureCube data upload

* [d3d11] Don't add bind flags render target for non RENDER_TARGET texture

Fix MotionStreak vertex format

* [d3d11] Fix blend state translation

* Fix lua bindings

* Adapte metal backend

* Fix metal backend

halx99 committed:
Remove OpenGL ES 2.0 support (#2633)

* Remove OpenGL ES 2.0 support

* Fix win32 dll link error [skip ci]

halx99 committed:
Rename renderer/backend to core/rhi

aismann commented:
Assertion failed!

Program: D:\_git\axmol\build\bin\cpp-tests\Debug\cpp-tests.exe
File: D:\_git\axmol\core\rhi\opengl\ProgramGL.cpp
Line: 143

Expression: vertShader != 0 && fragShader != 0

<img width="1117" height="164" alt="image" src="https://github.com/user-attachments/assets/b5e5c58a-5220-4f8b-bff8-568923307adc" />

halx99 commented:
> Assertion failed!
> 
> Program: D:_git\axmol\build\bin\cpp-tests\Debug\cpp-tests.exe File: D:_git\axmol\core\rhi\opengl\ProgramGL.cpp Line: 143
> 
> Expression: vertShader != 0 && fragShader != 0
> 
> <img alt="image" width="1016" height="108" src="https://private-user-images.githubusercontent.com/8652787/477996123-c9085985-c4f9-4241-83df-3681e977cbce.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTUxNzIzNTQsIm5iZiI6MTc1NTE3MjA1NCwicGF0aCI6Ii84NjUyNzg3LzQ3Nzk5NjEyMy1jOTA4NTk4NS1jNGY5LTQyNDEtODNkZi0zNjgxZTk3N2NiY2UucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDgxNCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTA4MTRUMTE0NzM0WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ODkwNGUyYzNiYzAyNjExODNiMDIzMDgwOWY4ZjhkYWFhM2NhZGVmY2JlYmUwNjgzZDAyODkzOThhODc3ODEyOSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ._9THt_XFmeiNVNSF7EvF15Opm8NRfBW7nDPeIsY1q8k">

I run on my local computer, works fine, you should `pull --rebase`, and do cmake regenerate

aismann commented:
> I run on my local computer, works fine, you should `pull --rebase`, and do cmake regenerate


It was a "fresh" cloned repository and I tried also the artifact from the build (RelWithDebInfo) which is like a "standalone installation"
Artifact download URL: https://github.com/axmolengine/axmol/actions/runs/16961100151/artifacts/3763459927

aismann commented:
Found the "error":
axsc => correct axslc
<img width="136" height="103" alt="image" src="https://github.com/user-attachments/assets/cf078b72-c545-4915-a153-62cf51b26065" />

halx99 commented:
> Found the "error": axsc => correct axslc <img alt="image" width="136" height="103" src="https://private-user-images.githubusercontent.com/8652787/478006340-cf078b72-c545-4915-a153-62cf51b26065.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTUxODI3NTUsIm5iZiI6MTc1NTE4MjQ1NSwicGF0aCI6Ii84NjUyNzg3LzQ3ODAwNjM0MC1jZjA3OGI3Mi1jNTQ1LTQ5MTUtYTE1My02MmNmNTFiMjYwNjUucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDgxNCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTA4MTRUMTQ0MDU1WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NGJlNjEyOTJmYzczMThiOWQ5MGUyNzQwOTViOWM3NzhhODFkNmU1NzQ1ZTk4NDE2MTA4ZGVlOTgxM2ExN2RlYyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.qvqgr_69GpjifoW5cLPIuNPSrjbap4J1Bc7vIj9SsPI">

should fixed by latest push, please check, thanks

aismann commented:
> 
> should fixed by latest push, please check, thanks

Works. Thanks. And I like the 3.0!

halx99 committed:
Rename folder core => axmol (step.1)

halx99 committed:
Rename folder core => axmol (step.2)

halx99 committed:
Rename folder core => axmol (step.3)

halx99 committed:
Restrict vs_inst buffer to read-only usage in vertex shader

halx99 committed:
Fix lint warnings

halx99 committed:
[RHI] Unify several descriptor type names

halx99 committed:
[RHI] Add texture sampler cache support (#2639)

j-jorge commented:
I was wondering if v3 wouldn't be a good time for more breaking changes, like getting rid of the `.h` extension for the headers in favor of `.hpp` or something like this, such that we can finally know that it is C++ without opening the file. What do you think?

In the same vein, what do you think of enforcing the code formatting in the CI by running a clang-format check on each PR?

halx99 committed:
Merge branch 'dev' into wip-v3

halx99 committed:
[RHI] Add sampler2DArray support (#2641)

* [RHI] Add sampler2DArray support

* [RHI] Add sampler2DArray support, adpot mipmaps

* [RHI] Add sampler2DArray support, adpot RenderTexture

* [RHI] Add sampler2DArray support, adpot etc1 with alpha rendering

* [RHI] Add sampler2DArray support, adpot context loss recovery

* [RHI] Add sampler2DArray support, improve code style

halx99 committed:
Fix a regression introduced by commit 127ddac that cause parse plist sprite sheet incorrect

halx99 committed:
Fix progress timer render on d3d11 (#2644)

The d3d11 will not fill missing a_position comp z,w like opengl, so need matched shader to render 2d nodes: ProgressTimer, MotionStreak, Grid, Texture2D

halx99 committed:
[d3d11] Add DepthStencilState cache support (#2645)

halx99 committed:
Improve VertexLayout management (#2647)

* Improve VertexLayout management

* Unify instance draw shader

* Update d3d shader compile inital optimize flag

* comment vertex input binding cache [skip ci]

halx99 committed:
Re-add ProgramState setTextureArray feature and do improvement (#2649)

* Re-add ProgramState setTextureArray feature and do improvement

* OMG, the d3d11 terrain rendering works as expected

And  previous revision does't work as expected due to texture binding slots incorrect

* Restore `ProgramState::setTextureArray`
* Improve Terrain rendering, use `uniform sampler2D u_details[4]` to store detailMap textures
* Improve  Terrain rendering,  set alphaMap, lightMap dummy texture RHI independent when they are not used
* Fix Terrain rendering doesn't work when use d3d11
* Fix vertex attribute binding cache incorrect
* Use macos-14 for ios ci due to unexpected compile error
* Upgrade MSL version: 1.2 => 2.0 (requires macos13 and ios 11+) for supporting texture array `sampler2D u_details[4]`

halx99 committed:
[d3d11] Implement CommandBufferImpl::readPixels

halx99 commented:
This PR is for ci check only, since we try to update branch mangement: enable ci for dev/*, release/*, this PR can closed, v3 roadmap, please see: https://github.com/axmolengine/axmol/discussions/2650

halx99 commented:
> I was wondering if v3 wouldn't be a good time for more breaking changes, like getting rid of the `.h` extension for the headers in favor of `.hpp` or something like this, such that we can finally know that it is C++ without opening the file. What do you think?
> 
> In the same vein, what do you think of enforcing the code formatting in the CI by running a clang-format check on each PR?

auto run clang-foramt check is a good idea.

halx99 commented:
> > I was wondering if v3 wouldn't be a good time for more breaking changes, like getting rid of the `.h` extension for the headers in favor of `.hpp` or something like this, such that we can finally know that it is C++ without opening the file. What do you think?
> > In the same vein, what do you think of enforcing the code formatting in the CI by running a clang-format check on each PR?
> 
> auto run clang-foramt check is a good idea.

should finish by PR: #2654

j-jorge commented:
If I understood #2654 correctly it creates a new commit which formats the code on each PR. This is imho a bad solutions as those commits will pollute the history with meaningless changes, complexifying  `git blame` and `git bisect`, for example. Still imho, the pipeline should just check that the code is formatted, and let the committer rewrite its commits with adequately formatted code.

halx99 commented:
> If I understood #2654 correctly it creates a new commit which formats the code on each PR. This is imho a bad solutions as those commits will pollute the history with meaningless changes, complexifying `git blame` and `git bisect`, for example. Still imho, the pipeline should just check that the code is formatted, and let the committer rewrite its commits with adequately formatted code.

at least, first time, we can format all sources
